do the ORG filtering as part of the mongo query. Also, don't trigger met...

chrisndodge opened the pull request:
...adata inheritence calculations when just trying to load 'about' information, which is not tied into the course tree anyhow (i.e. no parent/child relationship)

Some code cleanup regarding eligibility to apply inherited metadata

forgot to return result

chrisndodge commented:
@cpennington I wrote this a long time ago, but abandoned it. Our homepage load on MIT PE has gotten unbearable and all current evidence is pointing to the fact that the ORG filtering is happening on the Django side as opposed to being part of the Mongo query.

The filter parameter is implemented for DraftModuleStore but not the SplitMongoStore version of the modulestore() interface implementation. What would it take to do a similar filtering on the Split side of things?

cpennington commented on common/lib/xmodule/xmodule/modulestore/mongo/base.py:
```diff
@@ -482,6 +482,8 @@ def do_connection(
 
         self._course_run_cache = {}
 
+        self.detached_categories = [name for name, __ in XBlock.load_tagged_classes("detached")]
```
You may as well make this a global constant, since the list is set at install-time.

cpennington commented:
The `course_index` in split has an `org` field, so you should be able to filter using that. I tend to agree with @dmitchell's suggestion to just call the filtering parameter `org`.

chrisndodge commented:
Looking at the split code, is this where I should add an optional filter?

    def _get_structures_for_branch_and_locator(self, branch, locator_factory, **kwargs):

        """
        Internal generator for fetching lists of courses, libraries, etc.
        :param str branch: Branch to fetch structures from
        :param type locator_factory: Factory to create locator from structure info and branch
        """
        result = []
        for entry, structure_info in self._get_structures_for_branch(branch):
            locator = locator_factory(structure_info, branch)
            <-- apply filter here on locator.org -->
            envelope = CourseEnvelope(locator, entry)
            root = entry['root']
            structures_list = self._load_items(envelope, [root], depth=0, **kwargs)
            if not isinstance(structures_list[0], ErrorDescriptor):
                result.append(structures_list[0])
        return result

cpennington commented:
Yeah, that looks like the right place (although you're going to want to plumb the filter into `_get_structures_for_branch`, and then `find_matching_course_indexes`, which is the thing that finally does the mongo query.

chrisndodge commented:
@cpennington OK, wired through filter on split.py and added unit tests for both draft and split. Does this look right?

Do you have a recommendation for a 2nd reviewer?

chrisndodge commented on common/lib/xmodule/xmodule/modulestore/split_mongo/mongo_connection.py:
```diff
@@ -208,6 +210,9 @@ def find_matching_course_indexes(self, branch=None, search_targets=None):
 
+        if org_target:
+            query['org'] = org_target
+
```
NOTE: I couldn't use the existing 'search_targets' parameters because that only filters arbitrary keys in a 'search_targets_' sub-dict in the mongo document. 'org' is at the root of the document.

cpennington replied:
Yeah, `search_targets` is intended for particular content that has been pulled up to the root for querying.

chrisndodge commented:
@cpennington everything passes now. Can I get a formal review so we can deploy in next week's release at the latest?

Also, who would be a good 2nd reviewer?

Thank you

cpennington commented:
@doctoryes, can you review this?

cpennington commented on common/lib/xmodule/xmodule/modulestore/mongo/base.py:
```diff
@@ -933,17 +935,36 @@ def _load_items(self, course_key, items, depth=0, using_descriptor_system=None):
         '''
-        Returns a list of course descriptors.
+        Returns a list of course descriptors. This accepts an optional parameter of 'org' which
+        will apply an efficient filter to only get courses with the specified ORG
```
Optional: It might be nice to have `org` as an explicit keyword arg in the function signature, since you've made it something supported by all modulestores.

cpennington commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -486,14 +486,16 @@ def version_block(self, block_data, user_id, update_version):
+        indexes = self.db_connection.find_matching_course_indexes(branch, search_targets, org_target)
 
         def _replace_or_append_index(altered_index):
             """
```
You need to add filtering for `org` to the section of this method that handles active `bulk_op` records. (Line 513 and below)

chrisndodge replied:
OK, I think I got this and wrote a unit test to cover it (I think)

cpennington commented on common/lib/xmodule/xmodule/modulestore/tests/test_mongo.py:
```diff
@@ -4,6 +4,7 @@
 # pylint: disable=no-member
 # pylint: disable=protected-access
 # pylint: disable=no-name-in-module
+# pylint: disable=bad-continuation
```
It would be better to just fix the indentation, I think.

chrisndodge replied:
It was strange pylint check on my localdev was passing, but build on Jenkins wasn't. I don't know if there is a pylint version skew or not. That said, I couldn't figure out what the problem was about (around line 213 of test_mongo.py - and similar block right below it)

cpennington commented:
Once my comments are addressed, :+1:

chrisndodge commented:
@doctoryes can you also do a pass? We'd like to get this in next week's release. TBH, I'm not sure why Jenkin's build is having a problem with the indentation on test_mongo.py:211'ish but my localdev run of code quality is not complaining of the indentation rules...

doctoryes commented on common/lib/xmodule/xmodule/modulestore/tests/test_mongo.py:
```diff
@@ -212,6 +228,39 @@ def test_get_courses(self):
+            assert_in(course_key, course_ids)  # pylint: disable=no-value-for-parameter
+
+        courses = self.draft_store.get_courses(org='edX')
+        assert_equals(len(courses), 6)  # pylint: disable=no-value-for-parameter
```
Why is this assert check for 6 courses - and then five courses are searched for below? Where's the sixth course below?

doctoryes commented on common/lib/xmodule/xmodule/modulestore/tests/test_mongo.py:
```diff
@@ -191,15 +204,18 @@ def test_mongo_modulestore_type(self):
         '''Make sure the course objects loaded properly'''
         courses = self.draft_store.get_courses()
-        assert_equals(len(courses), 6)
+        assert_equals(len(courses), 7)  # pylint: disable=no-value-for-parameter
```
Why are there seven courses above - and only six explicitly checked against below?

chrisndodge replied:
Hrm. This needs more investigation. If I run this specific test there are 6, if I run the whole suite then there are 7. Seems like probably another test run isn't cleaning something up in the mongo database....

doctoryes commented:
Looks good - besides questions about tests, :+1:

chrisndodge commented:
@doctoryes @cpennington I added a bunch of documentation to that unit_test which is returning the cached courses. I tried flushing cache between unit tests, but didn't work (is it using Django cache? maybe a special cache name rather than the default?). I'm going to close this PR and open a new one onto the Release Candidate and then merge.

FYI @mattdrayer

doctoryes commented:
JIRA ticket for the test:

https://openedx.atlassian.net/browse/PLAT-468
